### PR TITLE
Shibboleth Auth Provider + LDAP Search

### PR DESCRIPTION
**Pull request by luthermonson**
Adds a new SAML auth provider for Shibboleth but couples it with OpenLDAP searching for Principals. It operates as a standalone SAML provider with no needed extra configs but if you backend Shibboleth with OpenLDAP you can configure it like a regular OpenLDAP auth provider but only works for get/search so you can find and assign principals to Clusters/Projects/etc.

https://github.com/rancher/rancher/issues/24235
requires: https://github.com/rancher/types/pull/1080

### CODE REVIEW

File: pkg/auth/providers/ldap/ldap_provider.go [hunk 1]
```diff
@@ -141,6 +145,53 @@ func (p *ldapProvider) GetPrincipal(principalID string, token v3.Token) (v3.Prin
 		return v3.Principal{}, err
 	}
 
+	if p.samlSearchProvider() {
+		config, caPool, err := p.getLDAPConfig()
```
Comment (mrajashree) [inline pkg/auth/providers/ldap/ldap_provider.go]: we don't need to call getLDAPConfig here again since it's called on line 138

File: pkg/auth/providers/saml/saml_provider.go [hunk 2]
```diff
@@ -243,22 +334,24 @@ func (s *Provider) SearchPrincipals(searchKey, principalType string, token v3.To
...
 		return v3.Principal{}, fmt.Errorf("SAML: invalid id %v", principalID)
 	}
-	principalType := parts[0]
-	externalID := strings.TrimPrefix(parts[1], "//")
-
 	if principalType != s.userType && principalType != s.groupType {
 		return v3.Principal{}, fmt.Errorf("SAML: Invalid principal type")
 	}
 
+	logrus.Debugf("GET PRINCIPAL %s \n", principalID)
```
Comment (mrajashree) [inline pkg/auth/providers/saml/saml_provider.go]: nit: is this in all caps by mistake? :)

File: pkg/auth/providers/ldap/ldap_client.go [hunk 3]
```diff
@@ -521,6 +549,5 @@ func (p *ldapProvider) RefetchGroupPrincipals(principalID string, secret string)
 	if err != nil {
```
Comment (mrajashree) [inline pkg/auth/providers/ldap/ldap_client.go]: Since this will generate group principals using ldap's attributes, we'll need to change them according to the saml ones just like you're doing for searchPrincipals

File: pkg/auth/providers/ldap/ldap_client.go [hunk 4]
```diff
@@ -423,8 +433,33 @@ func (p *ldapProvider) searchLdap(query string, scope string, config *v3.LdapCon
...
+		externalId := results.Entries[i].DN
 		entry := results.Entries[i]
-		principal, err := ldap.AttributesToPrincipal(entry.Attributes, results.Entries[i].DN, scope, p.providerName, config.UserObjectClass, config.UserNameAttribute, config.UserLoginAttribute, config.GroupObjectClass, config.GroupNameAttribute)
+
+		if p.samlSearchProvider() {
+			if strings.EqualFold("user", entityType) {
+				userLoginValues := ldap.GetAttributeValuesByName(entry.Attributes, config.UserLoginAttribute)
+				if len(userLoginValues) > 0 {
+					externalId = userLoginValues[0] // only support first
+				}
```
Comment (cjellick) [inline pkg/auth/providers/ldap/ldap_client.go]: so for user lookup to work properly, the user must ensure that the value of the userLoginAttribute field is the same as some value configured in the saml config, right?

File: pkg/auth/providers/ldap/ldap_provider.go [hunk 5]
```diff
@@ -141,6 +145,46 @@ func (p *ldapProvider) GetPrincipal(principalID string, token v3.Token) (v3.Prin
...
 
+	if p.samlSearchProvider() {
+		lConn, err := ldap.Connect(config, caPool)
+		if err != nil {
+			return v3.Principal{}, nil
+		}
+		defer lConn.Close()
+
+		if scope == p.userScope {
+			userPrincipals, err := p.searchUser(externalID, config, lConn)
```
Comment (cjellick) [inline pkg/auth/providers/ldap/ldap_provider.go]: I dont think leveraging searchUser is really the right thing to be doing here. That function is for fuzzy searching, not for hard lookups, which this is. You should use something similar to the look up we do as part of login:

https://github.com/rancher/rancher/blob/7330335a207c6c0ab98d0b8606a5fdf8cde5bd12/pkg/auth/providers/ldap/ldap_client.go#L46-L49

File: pkg/auth/providers/ldap/ldap_provider.go [hunk 6]
```diff
@@ -141,6 +145,46 @@ func (p *ldapProvider) GetPrincipal(principalID string, token v3.Token) (v3.Prin
...
+			principal := &userPrincipals[0]
+			if p.isThisUserMe(token.UserPrincipal, *principal) {
+				principal.Me = true
+			}
+
+			return *principal, nil
+		}
+
+		if scope == p.groupScope {
+			groupPrincipals, err := p.searchGroup(externalID, config, lConn)
```
Comment (cjellick) [inline pkg/auth/providers/ldap/ldap_provider.go]: Same as above. You are doing a hard lookup, not a fuzzy search. Should probably have a more explicit query.

File: pkg/auth/providers/saml/saml_provider.go [hunk 7]
```diff
@@ -31,21 +34,26 @@ const (
...
+	ctx                context.Context
+	authConfigs        v3.AuthConfigInterface
+	secrets            corev1.SecretInterface
+	userMGR            user.Manager
+	tokenMGR           *tokens.Manager
+	serviceProvider    *saml.ServiceProvider
+	name               string
+	userType           string
+	groupType          string
+	certs              string
```
Comment (cjellick) [inline pkg/auth/providers/saml/saml_provider.go]: anything that is ldap specific, prefix it as such. But also, why arent the certs and caPool part of ldapProvider?

File: pkg/auth/providers/saml/saml_provider.go [hunk 8]
```diff
@@ -120,6 +145,52 @@ func PerformSamlLogin(name string, apiContext *types.APIContext, input interface
 	return nil
 }
 
+func (s *Provider) getLDAPConfig() (*v3.LdapConfig, *x509.CertPool, error) {
```
Comment (cjellick) [inline pkg/auth/providers/saml/saml_provider.go]: Is this code just all copy and pasted from somewhere else? If yes, why did you need to do that rather than reuse the existing function (if there is no easy way to reuse, thats fine, just want to know).

File: pkg/auth/providers/saml/saml_provider.go [hunk 9]
```diff
@@ -223,13 +300,27 @@ func (s *Provider) toPrincipal(principalType string, princ v3.Principal, token *
 }
 
 func (s *Provider) RefetchGroupPrincipals(principalID string, secret string) ([]v3.Principal, error) {
-	// This should never be called
+	if s.hasValidLdapGroupSearch() {
```
Comment (cjellick) [inline pkg/auth/providers/saml/saml_provider.go]: as discussed, removed refetch for saml
